azure/services/virtualnetworks/virtualnetworks.go

```diff
@@ -18,151 +18,130 @@ package virtualnetworks
 	ctx, log, done := tele.StartSpanWithLogger(ctx, "virtualnetworks.Service.Delete")
 	defer done()
 
-	vnetSpec := s.Scope.VNetSpec()
-	existingVnet, err := s.getExisting(ctx, vnetSpec)
-	if azure.ResourceNotFound(err) {
-		// vnet does not exist, there is nothing to delete
-		return nil
-	}
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
 
-	if !existingVnet.IsManaged(s.Scope.ClusterName()) {
-		log.V(4).Info("Skipping VNet deletion in custom vnet mode")
-		return nil
-	}
+	vnetSpec := s.Scope.VNetSpec()
 
-	log.V(2).Info("deleting VNet", "VNet", vnetSpec.Name)
-	err = s.Client.Delete(ctx, vnetSpec.ResourceGroup, vnetSpec.Name)
+	// Check that the vnet is not BYO.
+	managed, err := s.IsManaged(ctx, vnetSpec)
 	if err != nil {
-		if azure.ResourceGroupNotFound(err) || azure.ResourceNotFound(err) {
+		if azure.ResourceNotFound(err) {
+			// already deleted or doesn't exist, cleanup status and return.
+			s.Scope.DeleteLongRunningOperationState(vnetSpec.ResourceName(), serviceName)
+			s.Scope.UpdateDeleteStatus(infrav1.VNetReadyCondition, serviceName, nil)
 			return nil
 		}
+		return errors.Wrap(err, "could not get VNet management state")
 	}
-	if err != nil {
-		return errors.Wrapf(err, "failed to delete VNet %s in resource group %s", vnetSpec.Name, vnetSpec.ResourceGroup)
+	if !managed {
+		log.Info("Skipping VNet deletion in custom vnet mode")
+		return nil
 	}
 
-	log.V(2).Info("successfully deleted VNet", "VNet", vnetSpec.Name)
-	return nil
+	err = s.DeleteResource(ctx, vnetSpec, serviceName)
+	s.Scope.UpdateDeleteStatus(infrav1.VNetReadyCondition, serviceName, err)
+	return err
 }
 
-// getExisting provides information about an existing virtual network.
-func (s *Service) getExisting(ctx context.Context, spec azure.VNetSpec) (*infrav1.VnetSpec, error) {
-	ctx, log, done := tele.StartSpanWithLogger(ctx, "virtualnetworks.Service.getExisting")
+// IsManaged returns true if the virtual network has an owned tag with the cluster name as value,
+// meaning that the vnet's lifecycle is managed, and caches the result in scope so that other services that depend on the vnet can check if it is managed.
+func (s *Service) IsManaged(ctx context.Context, spec azure.ResourceSpecGetter) (bool, error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "virtualnetworks.Service.IsManaged")
 	defer done()
 
-	vnet, err := s.Client.Get(ctx, spec.ResourceGroup, spec.Name)
+	if spec == nil {
+		return false, errors.New("cannot get vnet to check if it is managed: spec is nil")
+	}
+
+	vnetIface, err := s.Get(ctx, spec)
 	if err != nil {
 		if azure.ResourceNotFound(err) {
-			log.V(2).Info(fmt.Sprintf("Resource not found for VNet %q from resource group %q", spec.Name, spec.ResourceGroup))
-			return nil, err
+			// if the vnet was already deleted, attempt to get previous management state from the spec.
+			if s.Scope.Vnet().ID != "" {
+				return s.Scope.Vnet().Tags.HasOwned(s.Scope.ClusterName()), nil
```

Comment:
why should we do this? can we not just return the error since `azure.ResourceNotFound` is handled in the calling function?

Reply:
I think I did this at the time I was sharing this function with other services trying to centralize "is vnet managed" because there was an edge case where if the vnet had already been deleted we wouldn't know whether we should delete the nat gateway. I don't think we need this for now though since I removed the caching part of this PR to simplify it. Will remove and update comment.

Reply:
done